gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java

```diff
@@ -748,4 +766,26 @@ private List<Factor> filteredFactorCatalog(List<Factor> factors){
                 .filter(factor -> !FactorType.RECOVERY_CODE.equals(factor.getFactorType()))
                 .collect(Collectors.toList());
     }
+
+    private void updateAuditLog(String type, User endUser, Client client, Factor factor, EnrolledFactor enrolledFactor, Throwable cause) {
+        final EnrolledFactorChannel channel = enrolledFactor.getChannel();
+
+        final MFAAuditBuilder builder = AuditBuilder.builder(MFAAuditBuilder.class)
+                .user(endUser)
+                .factor(factor)
+                .type(type)
+                .channel(channel);
+
+        if (client != null) {
```

Comment:
Could some logic be moved to Builder: ifs, enrolledFactor.getChannel() instead of duplicating it in two places?

Reply:
Avoiding duplication won't be possible easily without refactoring few packages.

Reply:
ifs are moved to Builder and it is fine for me now